code/datums/mapgen/CaveGenerator.dm

```diff
@@ -106,7 +125,95 @@
 	to_chat(world, span_boldannounce("[message]"))
 	log_world(message)
 
+
+/**
+ * This proc handles including biomes in the cave generation. This is slower than
+ * `generate_terrain()`, so please use it only if you actually need biomes.
+ *
+ * This should only be called by `generate_terrain()`, if you have to call this,
+ * you're probably doing something wrong.
+ */
+/datum/map_generator/cave_generator/proc/generate_terrain_with_biomes(list/turfs, area/generate_in)
+	if(!(generate_in.area_flags & CAVES_ALLOWED))
+		return
+
+	var/humidity_seed = rand(0, 50000)
+	var/heat_seed = rand(0, 50000)
+
+	var/start_time = REALTIMEOFDAY
+	string_gen = rustg_cnoise_generate("[initial_closed_chance]", "[smoothing_iterations]", "[birth_limit]", "[death_limit]", "[world.maxx]", "[world.maxy]") //Generate the raw CA data
+
+	var/list/open_turfs_used = list()
+
+	for(var/turf/gen_turf as anything in turfs) //Go through all the turfs and generate them
+		var/closed = string_gen[world.maxx * (gen_turf.y - 1) + gen_turf.x] != "0"
+		var/turf/new_turf = pick(closed ? closed_turf_types : open_turf_types)
```

Comment:
these datum var reads should be cached to the proc

Reply:
Which ones?

Reply:
closed_turf_types and open_turf_types
also we can cache world.maxx but that's a bit more nit picky